packages/resource-deleter/src/main.js

```diff
@@ -0,0 +1,185 @@
+  resource: string
+
+  constructor(options: resourceDeleterOptions) {
+    if (!options.apiConfig)
+      throw new Error('The constructor must passed an `apiConfig` object')
+
+    if (!options.resource) throw new Error('A `resource` object must be passed')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          options.accessToken ? `Bearer ${options.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow({
+          ...this.apiConfig,
+          fetch,
+        }),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl, fetch }),
+        createQueueMiddleware({
+          concurrency: 20,
+        }),
+      ],
+    })
+    this.logger = {
+      ...silentLogger,
+      ...options.logger,
+    }
+    this.predicate = options.predicate
+    this.resource = options.resource
+  }
+
+  run(): Promise<any> {
+    this.logger.info(`Starting to delete fetched ${this.resource}`)
+    const uri = this.buildUri(this.apiConfig.projectKey, this.predicate)
+    const request = ResourceDeleter.buildRequest(uri, 'GET')
+
+    return this.client
+      .process(
+        request,
+        (response: ClientResponse): Promise<any> => {
+          if (response.statusCode !== 200) {
+            return Promise.reject(
+              new Error(`Request returned status code ${response.statusCode}`)
+            )
+          }
+          let results = response.body?.results
+
+          // Check if the resource is empty
+          if (!results || !results.length) {
+            this.logger.info(
+              `No ${this.resource} is found in the project ${
+                this.apiConfig.projectKey
+              }, therefore nothing to delete.`
+            )
+            return Promise.resolve('nothing to delete')
+          }
+
+          // Check if the resource is categories
+          if (this.resource === 'categories')
+            // Sort categories by their ancestors attributes sizes
+            results = results.sort(ResourceDeleter.sortCategories)
+
+          let deletedItems = 0
+          return Promise.all(
+            results.map(
+              async (result: Object): Promise<any> => {
+                let newVersion
+                // Check if the resource is published
+                if (result.masterData?.published) {
+                  await this.unPublishResource(result)
+                  newVersion = result.version + 1
+                }
+                deletedItems += 1
+                return this.deleteResource({
+                  ...result,
+                  version: newVersion || result.version,
+                })
+              }
+            )
+          ).then(
+            (): void =>
+              this.logger.info(`${deletedItems} ${this.resource} deleted`)
```

Comment:
Thank you for adding the counter but here it is pointless - because `deletedItems` will be every time equal to `results.length` (you always delete all loaded results). I thinking more about creating a global counter which would at the very end (line 122) log the total count of all items deleted (and not only count of deleted items in every batch).